lang/funcs/collection.go

```diff
@@ -497,7 +502,23 @@ var LookupFunc = function.New(&function.Spec{
 			return cty.NilType, fmt.Errorf("lookup() takes two or three arguments, got %d", len(args))
 		}
 
-		return args[0].Type().ElementType(), nil
+		ty := args[0].Type()
+		key := args[1].AsString()
+		switch {
+		case ty.IsObjectType():
+			if ty.HasAttribute(key) {
+				return args[0].GetAttr(key).Type(), nil
+			} else if len(args) == 3 {
+				// if the key isn't found but a default is provided,
+				// return the default type
+				return args[2].Type(), nil
+			}
+			return cty.DynamicPseudoType, fmt.Errorf("the given object has no attribute %q", key)
+		case ty.IsMapType():
+			return ty.ElementType(), nil
+		default:
+			return cty.NilType, fmt.Errorf("lookup() requires a map as the first argument")
```

Comment:
Sorry I didn't think of this on the first go here but if we use `function.NewArgErrorf` here instead of `fmt.Errorf` then we can indicate the first argument in a machine-readable way, which can then allow the HCL layer to highlight the specific expression in the error message. (This applies also to the "given object has no attribute ..." error above)